src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs

```diff
@@ -0,0 +1,846 @@
+
+            public CompletionListUpdater(
+                IAsyncCompletionSession session,
+                AsyncCompletionSessionDataSnapshot data,
+                RecentItemsManager recentItemsManager,
+                IGlobalOptionService globalOptions,
+                CancellationToken cancellationToken)
+            {
+                _session = session;
+                _data = data;
+                _recentItemsManager = recentItemsManager;
+                _globalOptions = globalOptions;
+                _cancellationToken = cancellationToken;
+
+                _filterText = _session.ApplicableToSpan.GetText(_data.Snapshot);
+                _initialRoslynTriggerKind = Helpers.GetRoslynTriggerKind(_data.InitialTrigger);
+
+                if (!_session.Properties.TryGetProperty(CompletionSource.HasSuggestionItemOptions, out bool hasSuggestedItemOptions))
+                {
+                    // This is the scenario when the session is created out of Roslyn, in some other provider, e.g. in Debugger.
+                    // For now, the default hasSuggestedItemOptions is false.
+                    hasSuggestedItemOptions = false;
+                }
+
+                _hasSuggestedItemOptions = hasSuggestedItemOptions || _data.DisplaySuggestionItem;
+
+                // We prefer using the original snapshot, which should always be available from items provided by Roslyn's CompletionSource.
+                // Only use data.Snapshot in the theoretically possible but rare case when all items we are handling are from some non-Roslyn CompletionSource.
+                var snapshotForDocument = TryGetInitialTriggerLocation(_data, out var intialTriggerLocation)
+                    ? intialTriggerLocation.Snapshot
+                    : _data.Snapshot;
+
+                _document = snapshotForDocument?.TextBuffer.AsTextContainer().GetOpenDocumentInCurrentContext();
+                _completionService = _document?.GetLanguageService<CompletionService>();
+                _completionRules = _completionService?.GetRules(CompletionOptions.From(_document!.Project)) ?? CompletionRules.Default;
+
+                // Let us make the completion Helper used for non-Roslyn items case-sensitive.
+                // We can change this if get requests from partner teams.
+                _completionHelper = _document != null ? CompletionHelper.GetHelper(_document) : new CompletionHelper(isCaseSensitive: true);
+
+                // Nothing to highlight if user hasn't typed anything yet.
+                _highlightMatchingPortions = _filterText.Length > 0
+                    && _globalOptions.GetOption(CompletionViewOptions.HighlightMatchingPortionsOfCompletionListItems, _document?.Project.Language);
+
+            }
+
+            public FilteredCompletionModel? UpdateCompletionList()
+            {
+                if (ShouldDismissCompletionListImmediately())
```

Comment:
consdier doc'ing (with a potential example).

Reply:
it's also ok if you don't *as long as* the method thsi calls is suitably clear and docs its cases.  so for this, i'm ok with no docs.

Reply:
I think the comment inside `ShouldDismissCompletionListImmediately` has enough details

Reply:
sgtm